Added the blog (#1)
ayush2390/apify-blog: blog.mdx

# Building a Netflix Web Show Recommender with Crawlee

> Scraping data from websites Netflix is not as easy as scraping data from any regular website. Netflix's website heavily [depends on JavaScript](https://medium.com/dev-channel/a-netflix-web-performance-case-study-c0bcde26a9d9), and it consists of tons of elements. Crawlee has the capability of scraping JS-heavy websites using its headless browsers, which we will discuss further in the article

Author: Done

> 

Author: Done

> 

Reviewer: One line won't be enough I'm afraid. Please [read this guide](https://apify.notion.site/How-to-write-technical-tutorials-701d702523a24d72a889dfc513e0981d?pvs=4) and make sure that the end project is clearly shown. It's what the readers want to see first, before deciding whether they will spend time with the article.

> Crawlee supports autoscaling, headless browser support, and efficient proxy management that helps make scraping easy.

Author: Done

We will set up Crawlee and understand it's working. Let’s get started!

## Prerequisites

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Node.js from their [official website](https://nodejs.org/en/).

## Creating React app

Firstly, we will create a React app (for frontend) using Vite. You can check out the [ReactJS Docs](https://react.dev/) to create a React app.

> 

Author: Done

Once the React app is created, open it in VS Code.

![react](assets/react.png)

This will be the structure of your React app.

Run `npm run dev` command in the terminal to run the app.

![viteandreact](assets/viteandreact.png)

This will be the output displayed by the React app.

## Adding Scraper code

As per our project requirements, we will scrape the genres and the titles of the shows available on Netflix.

Let’s start building the scraper code.

### Installation

We will install two libraries - `crawlee` and `playwright`. Run this command to install these libraries:

```
npm install crawlee playwright
```

> `playwright` is not bundled with Crawlee to reduce install size and allow greater flexibility. Therefore we need to install both libraries separately.

Author: Done

Crawlee, built on Playwright, empowers [headless browser scraping](https://crawlee.dev/blog/scrapy-vs-crawlee#headless-browsing-and-js-rendering) for complex websites. While slower and more resource-intensive, it tackles intricate JavaScript-heavy scenarios that basic crawlers can't handle.

After installing the libraries, it’s time to create the scraper code.

### Importing libraries

`PlaywrightCrawler` class has been imported from the `crawlee` library. `PlaywrightCrawler` crawls the web using a headless browser controlled by the `Playwright` library.

> 

Author: Done

      if (allShows.length == 40) {
        totalShows.push(allShows);
        allShows = [];
      }
      allShows.push(show);
    });

    await pushData({
      genre: titles,
      shows: totalShows,
    });
  },

  // Let's limit our crawls to make our tests shorter and safer.
  maxRequestsPerCrawl: 20,
});

await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);
```

The `maxRequestPerCrawl` limits the crawlee to crawl only the first page of the given URL to make the tests shorter and safer. You can ignore this code line to crawl the entire website.

```js
await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);
```

In this code snippet, we have defined the URL that we want to scrape data from. You can scrape multiple URLs also by adding the other URLs in the array.

Now we will run the crawlee to scrape the website. Before running the crawlee, we need to tweak the `package.json` file. We will add the `start` script targeting the `scraper.js` file to run the crawlee.

Add the following code in `"scripts"` object:

```
"start": "node src/scraper.js"
```

and save it. Now run this command to run the crawlee the scrape the data:

> ```

Author: It is `npm start`. I have tried it

> ```

Reviewer: both will work. `npm start` is a shorthand for `npm run start`

npm start
```

After running this command, you will see a `storage` folder having `datasets\default\000000001.json` file. In this json file, the scrapped data will be stored in json format.

Now we can use this json data and display it in the `App.jsx` component to create the project.

> In the `App.jsx` component, you can see we have imported `jsonData` from the `000000001.json` file:

Author: Done

```js
import jsonData from "../storage/datasets/default/000000001.json";
```

Then we have used this `jsonData` and performed some `map` methods to get the desired results. Here is the full code of `App.jsx`:

```js
import { useState } from "react";
import "./App.css";
import jsonData from "../storage/datasets/default/000000001.json";

function App() {
  const [count, setCount] = useState();
  const handleChange = (event) => {
    setCount(event.target.value);
  };

  return (
    <>
      <h1>Netflix Web Show recommender</h1>
      <div>
        <select onChange={handleChange}>
          <option value="">Select your genre</option>;
          {jsonData.genre.map((genre, key) => {
            return <option value={key}>{genre}</option>;
          })}
        </select>
      </div>
      <div style={{ display: "flex" }} className="lists">

>         <div style={{ flex: "50%", paddingRight: "10px", textAlign: "left" }}>

Reviewer 2: UI doesn't look like the same from the code

Author: Done

        </div>
      </div>
    </>
  );
}

export default App;
```

In this code snippet, the `genre` array is used to display the list of genres. User can select their desired genre and based upon that a list of web shows available on Netflix will be displayed using the `shows` array.

Now our project is ready!

## Result

Now, to run your project in the localhost, run this command:

```
npm run dev
```

This command will run your project on the localhost. Here is the demo of the project:

![result](assets/result.gif)

Project link - [https://github.com/ayush2390/web-show-recommender](https://github.com/ayush2390/web-show-recommender)

In this way, you can scrape data from any website using crawlee and create some unique projects using the scraped data.

Want to know more about Crawlee? You can watch our Crawlee guide video:

<iframe
  width="560"
  height="315"
  src="https://www.youtube.com/embed/g1Ll9OlFwEQ?si=EMeYil29LV0oDOxc"
  title="YouTube video player"
  frameborder="0"
  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
  referrerpolicy="strict-origin-when-cross-origin"
  allowfullscreen

> ></iframe>

Reviewer 2: remove

Author: This will show the video tutorial from Apify Youtube channel. That's why I have added it

# Building a Netflix Web Show Recommender with Crawlee

> Scraping data from websites Netflix is not as easy as scraping data from any regular website. Netflix website heavily depends on JavaScript and it consists tons of elements. But Crawlee makes it super easy to scrape such complex websites.

Author: Done

> [Crawlee](https://crawlee.dev/), a web scraping and browser automation library, helps us crawl websites, scrape data from them, and store the scraped data at the defined location.

Author: Done

> Crawlee's features—autoscaling, proxy rotation, persistent URL queue, headless browser support, and Docker compatibility—make website scraping super easy.

Author: Done

In this blog, we will use Crawlee to scrape data from the Netflix website and use the scraped data to build a Netflix Web Show Recommender project.

We will set up Crawlee and understand how it works. Let’s get started.

## Prerequisites

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Nodejs from their [official website](https://nodejs.org/en/).

## Creating the project

To build the Web Show Recommender project, we will scrape data from the Netflix website, save it, and use it to build the project.

### Creating React app

Firstly, we will create a React app using Vite. You can check out the [ReactJS Docs](https://react.dev/) to create a React app.

Once the React app is created, open it in VS Code.

[img](assets/1.png)

> This will be the structure of your React app.

Author: Done

Run `npm run dev` command in the terminal to run the app.

This will be the output displayed by the React app.

### Adding Scraper code

Now, we will integrate Crawlee into our project and scrape the Netflix data using it.

As per our project requirements, we will scrape the genres and the titles of the shows available on Netflix.

Let’s start building the scraper code.

##### Installation

> We have to install two libraries - `crawlee` and `playwright`. Run this command to install these libraries:

Author: Done

```
npm install crawlee playwright
```

> [Crawlee](https://crawlee.dev/) is a web scraping and browser automation library. It helps to build crawlers to scrape the data from websites.

Author: Done

> [Playwright](https://github.com/microsoft/playwright) is a framework for Web Testing and Automation. It allows testing Chromium, Firefox, and WebKit with a single API. Created by Microsoft.

Author: Done

Crawlee, built on Playwright, empowers headless browser scraping for complex websites. While slower and more resource-intensive, it tackles intricate JavaScript-heavy scenarios that basic crawlers can't handle.

> 

Author: Done

After installing the libraries, it’s time to create the scraper code.

##### Importing libraries

> PlaywrightCrawler class has been imported from the `crawlee` library. `PlaywrightCrawler` crawls the web using a headless browser controlled by the `Playwright` library.

Author: Done

```js
import { PlaywrightCrawler } from "crawlee";
```

#### Scraping genres

To scrape the genres, we will utilize the [browser Devtools](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Tools_and_setup/What_are_browser_developer_tools) to identify the tags and CSS selectors targeting the genre elements on the Netflix website.

> Here, we can observe that the name of the genre is captured by the _span_ tag having _nm-collections-row-name_ class. So we can use the _span.nm-collections-row-name_ selector to capture this and similar elements.

Reviewer 2: ```suggestion
Here, we can see that the name of the genre is captured by the _span_ tag having `nm-collections-row-name` class. So we can use the `span.nm-collections-row-name` selector to capture this and similar elements.
```

Reviewer 2: change all of _ please, i can't mention each of it

```js
const titles = await page.$$eval("span.nm-collections-row-name", (elements) => {
  // Map the text content of each element
  return elements.map((el) => el.textContent.trim());
});
```

> This will give us the list of all the genres stored in _titles_ array.

Author: Done

> This will give us the list of all the genres stored in _titles_ array.

Reviewer 2: can't see any change

#### Scraping shows

To scrape the show titles, we'll use the same method we used for scraping genres. Let’s find out the tag capturing the shows’ names.

> Here, we can observe that the title of the show is captured by the _span_ tag having _nm-collections-title-name_ class. So we can use the _span.nm-collections-title-name_ selector to capture this and similar elements.

Author: Done

> Here, we can observe that the title of the show is captured by the _span_ tag having _nm-collections-title-name_ class. So we can use the _span.nm-collections-title-name_ selector to capture this and similar elements.

Reviewer 2: can't see any change

```js
const shows = await page.$$eval(
  "span.nm-collections-title-name",
  (elements) => {
    // Map the text content of each element
    return elements.map((el) => el.textContent.trim());
  }
);
```

> This will scrape the title of all the shows and store it in the _shows_ array.

Author: Done

> This will scrape the title of all the shows and store it in the _shows_ array.

Reviewer 2: no change here too

```js
let allShows = [];
let totalShows = [];
shows.map((show) => {
  if (allShows.length == 40) {
    totalShows.push(allShows);
    allShows = [];
  }
  allShows.push(show);
});
```

Here, we have performed some basic operations on the _shows_ array to get the _totalShows_ array in the desired format.

#### Storing data

Now we have every data that we want for our project and it’s time to store or save the scraped data. To store the data, crawlee comes with a **crawlee.pushData** method.

The [pushData()](https://crawlee.dev/docs/introduction/saving-data) method creates a storage folder in the project directory and stores the scraped data in the json format.

```js
await crawlee.pushData({
  genre: titles,
  shows: totalShows,

> });

Reviewer 2: crawlee.pushData?? what's that

Author: Done

```

This will save the titles and totalShows arrays as values in _genre_ and _shows_ keys.

Here’s the code that we will use in our project:

```js
import { PlaywrightCrawler } from "crawlee";

const crawler = new PlaywrightCrawler({
  requestHandler: async ({ page, request, enqueueLinks, pushData }) => {
    console.log(`Processing: ${request.url}`);
    console.log(request.label);

    const titles = await page.$$eval(
      "span.nm-collections-row-name",
      (elements) => {
        // Map the text content of each element
        return elements.map((el) => el.textContent.trim());
      }
    );
    // console.log(titles);
    const shows = await page.$$eval(
      "span.nm-collections-title-name",
      (elements) => {
        // Map the text content of each element
        return elements.map((el) => el.textContent.trim());
      }
    );

    let allShows = [];
    let totalShows = [];
    shows.map((show) => {
      if (allShows.length == 40) {
        totalShows.push(allShows);
        allShows = [];
      }
      allShows.push(show);
    });

    await pushData({

>       genre: titles,

Author: Done

# Building a Netflix Web Show Recommender with Crawlee

Scraping data from websites Netflix is not as easy as scraping data from any regular website. Netflix website heavily [depends on JavaScript](https://medium.com/dev-channel/a-netflix-web-performance-case-study-c0bcde26a9d9) and it consists tons of elements. Crawlee has the capability of scraping JS-heavy websites using its headless browsers, which we will discuss further

[Crawlee](https://crawlee.dev/) helps us crawl websites, scrape data from them, and store the scraped data at the defined location.

> Crawlee's features—autoscaling, persistent URL queue, headless browser support —make website scraping super easy.

Reviewer 2: again please write in a way that it not look like promotion.

> Crawlee supports autoscaling, headless browser support, and efficient proxy management that helps make scraping easy.

> In this blog, we will use Crawlee to scrape data from the Netflix website and use the scraped data to build a Netflix Web Show Recommender project.

Reviewer 2: the third mention of Crawlee and the project, why we need to say so many times the same thing?

We will set up Crawlee and understand how it works. Let’s get started.

## Prerequisites

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Nodejs from their [official website](https://nodejs.org/en/).

## Creating the project

> To build the Web Show Recommender project, we will scrape data from the Netflix website, save it, and use it to build the project.

Reviewer 2: 4th mention

### Creating React app

Firstly, we will create a React app (for frontend) using Vite. You can check out the [ReactJS Docs](https://react.dev/) to create a React app.

Once the React app is created, open it in VS Code.

![react](assets/react.png)

This will be the structure of your React app.

Run `npm run dev` command in the terminal to run the app.

![viteandreact](assets/viteandreact.png)

This will be the output displayed by the React app.

### Adding Scraper code

> Now, we will integrate Crawlee into our project and scrape the Netflix data using it.

Reviewer 2: 5th

# Building a Netflix Web Show Recommender with Crawlee and React

> Creating a web show recommender for Netflix involves scraping data from web pages efficiently. In this article, we'll guide you through the process of using Crawlee. By the end of this guide, you'll have a functional web show recommender that can fetch and scrape Netflix data to provide show suggestions. 

Reviewer: ```suggestion
In this blog, we'll guide you through the process of using Vite and Crawlee to build a website that recommends Netflix shows based on their categories and genres. To do that, we will first scrape the shows and categories from Neflix using Crawlee, and then visualize scraped data in a React app built with Vite. By the end of this guide, you'll have a functional web show recommender that can provide Netflix show suggestions. 
```

> Whether you're a developer looking to enhance your scraping skills or a Netflix enthusiast wanting to build a web-show recommendation engine, this project has something valuable for you. Let's dive in and explore how Crawlee can help us achieve this.
> 
> Crawlee supports autoscaling, headless browser support, and efficient proxy management that makes scraping easy.
> 
> We will set up Crawlee and understand it's working. Let’s get started!

Reviewer: I'd remove this. It has very little informational value for this tutorial.

```suggestion
Let’s get started!
```

## Prerequisites

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Node.js from their [official website](https://nodejs.org/en/).

## Creating React app

Firstly, we will create a React app (for frontend) using Vite. Run this command in the terminal to create react app:

```
npx create-vite@latest
```

You can check out the [Vite Docs](https://vitejs.dev/guide/) to create a React app.

Once the React app is created, open it in VS Code.

![react](assets/react.png)

This will be the structure of your React app.

Run `npm run dev` command in the terminal to run the app.

![viteandreact](assets/viteandreact.png)

This will be the output displayed by the React app.

## Adding Scraper code

As per our project requirements, we will scrape the genres and the titles of the shows available on Netflix.

Let’s start building the scraper code.

### Installation

We will install two libraries - `crawlee` and `cheerio`. Run this command to install these libraries:

```

> npm install crawlee cheerio

Reviewer: We don't need to install Cheerio separately.

```suggestion
npm install crawlee
```

```

> :::note
> `cheerio` is not bundled with Crawlee to reduce install size and allow greater flexibility. Therefore we need to install both libraries separately.
> :::

Reviewer: This isn't true. Deleting.

```suggestion
```

> Crawlee utilizes Cheerio for [HTML parsing and scraping](https://crawlee.dev/blog/scrapy-vs-crawlee#html-parsing-and-scraping) of static websites. While faster and [less resource-intensive](https://crawlee.dev/docs/guides/scaling-crawlers), it can only scrape websites that do not require JavaScript rendering, making it unsuitable for SPAs (single page applications).

Reviewer: This is just a rule of thumb, but usually it's better to link to API reference or guides in the documentation, because they get reviewed and updated. Blogs tend to get outdated.

> Additionally, Crawlee comes with the support of [Playwright](https://playwright.dev/) and [Puppeteer](https://pptr.dev/) for websites that are JavaScript-heavy in nature. 

Reviewer: ```suggestion
Additionally, Crawlee supports headless browser libraries like [Playwright](https://playwright.dev/) and [Puppeteer](https://pptr.dev/) for scraping of websites that are JavaScript-rendered.
```

After installing the libraries, it’s time to create the scraper code.

Create a file in `src` directory and name it `scraper.js`. Whole scraper code will be created in this file.

### Scraping genres

> To scrape the genres, we will utilize the [browser Devtools](https://developer.mozilla.org/en-US/docs/Learn/Common`questions/Tools`and`setup/What`are`browser`developer`tools) to identify the tags and CSS selectors targeting the genre elements on the Netflix website.

Reviewer: ```suggestion
To scrape the genres, we will utilize the [browser DevTools](https://developer.mozilla.org/en-US/docs/Learn/Common`questions/Tools`and`setup/What`are`browser`developer`tools) to identify the tags and CSS selectors targeting the genre elements on the Netflix website.
```

![genre](assets/genre.png)

Here, we can observe that the name of the genre is captured by the `span` tag having `nm-collections-row-name` class. So we can use the `span.nm-collections-row-name` selector to capture this and similar elements.

> ```js
> // Use parseWithCheerio for efficient HTML parsing
>     const $ = await parseWithCheerio();
> 
> // Extract genre titles
>     const titles = $(".nm-collections-row-name").map((_, el) => $(el).text().trim()).get();
> ```

Reviewer: Wrong indentation

```suggestion
```js
// Use parseWithCheerio for efficient HTML parsing
const $ = await parseWithCheerio();

// Extract genre titles
const titles = $(".nm-collections-row-name").map((_, el) => $(el).text().trim()).get();
```
```

This will give us the list of all the genres stored in the `titles` array.

### Scraping shows

To scrape the titles of shows, we will use the same method we used for scraping genres. Let’s find out the tag capturing the shows’ names.

![title](assets/title.png)

Here, we can observe that the title of the show is captured by the `span` tag having `nm-collections-title-name` class. So we can use the `span.nm-collections-title-name` selector to capture this and similar elements.

> ```js
> // Extract show titles
>     const shows = $(".nm-collections-title-name").map((_, el) => $(el).text().trim()).get();
> ```

Reviewer: ```suggestion
```js
// Extract show titles
const shows = $(".nm-collections-title-name").map((_, el) => $(el).text().trim()).get();
```
```

This will scrape the title of all the shows and store it in the `shows` array.

```js
// Prepare data for pushing
    const allShows = [];
    let chunk = [];
    shows.forEach((show) => {
      chunk.push(show);
      if (chunk.length === 40) {
        allShows.push(chunk);
        chunk = [];
      }
    });
    if (chunk.length > 0) {
      allShows.push(chunk);
    }

> ```

Reviewer: Why do all the code blocks have wrong indentation? Pls fix.

Reviewer: Btw, "prepare data for pushing" is not a super descriptive comment. It might be worth showing an example of the data structure we need to transform, and the needed result. This would visually explain why we need to do this chunking transformation. Also `chunk` is a very generic variable name. Can't we use something more descriptive? Why do we need to chunk it in increments of `40`?

> This code takes a large list of shows in `shows` and breaks them into groups of 40 in `chunk`, storing these groups in `allShows` array.

Reviewer: I can see this by looking at the code. But it does not explain WHY the code is doing that. Why do I need chunks of 40 in the first place?

### Storing data

Now we have every data that we want for our project and it’s time to store or save the scraped data. To store the data, crawlee comes with a `pushData()` method.

The [pushData()](https://crawlee.dev/docs/introduction/saving-data) method creates a storage folder in the project directory and stores the scraped data in the json format.

```js
await pushData({
      genre: titles,
      shows: allShows,
    });
```  

This will save the `titles` and `totalShows` arrays as values in the `genre` and `shows` keys.

Here’s the full code that we will use in our project:

```js

> import { CheerioCrawler, log, Dataset } from "crawlee";

Reviewer: We use single quotes in Crawlee. Please lint the code using Crawlee ESLint rules.

const crawler = new CheerioCrawler({
  requestHandler: async ({ request, parseWithCheerio, pushData }) => {
    log.info(`Processing: ${request.url}`);

    // Use parseWithCheerio for efficient HTML parsing
    const $ = await parseWithCheerio();

    // Extract genre titles
    const titles = $(".nm-collections-row-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Extract show titles
    const shows = $(".nm-collections-title-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Prepare data for pushing
    const allShows = [];
    let chunk = [];
    shows.forEach((show) => {
      chunk.push(show);
      if (chunk.length === 40) {
        allShows.push(chunk);
        chunk = [];
      }
    });
    if (chunk.length > 0) {
      allShows.push(chunk);
    }

    await pushData({
      genre: titles,
      shows: allShows,
    });
  },

  // Limit crawls for efficiency
  maxRequestsPerCrawl: 20,
});

> await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);

Reviewer: Why are we limiting the scrape to 20 requests, when we're only scraping a single URL? There's no enqueueing in the `requestHandler`

await Dataset.exportToJSON("results");
```

Now, we will run Crawlee to scrape the website. Before running Crawlee, we need to tweak the `package.json` file. We will add the `start` script targeting the `scraper.js` file to run Crawlee.

Add the following code in `"scripts"` object:

```
"start": "node src/scraper.js"
```

and save it. Now run this command to run the crawlee the scrape the data:

> ```

Reviewer: ```suggestion
```sh
```

npm start
```

After running this command, you will see a `storage` folder having `key_value_stores/default/results.json` file. In this json file, the scrapped data will be stored in json format.

Now we can use this json data and display it in the `App.jsx` component to create the project.

In the `App.jsx` component, we will import `jsonData` from the `results.json` file:

```js
import jsonData from "../storage/key_value_stores/default/results.json";
```

Then we have used this `jsonData` and performed some `map` methods to get the desired results. Here is the full code of `App.jsx`:

```js
import { useState } from "react";
import "./App.css";
import jsonData from "../storage/key_value_stores/default/results.json";

function App() {
  const [count, setCount] = useState(null);

  const handleChange = (event) => {

>     const value = event.target.value;
>     setCount(value === "" ? null : parseInt(value));
>   };

Reviewer: This seems mor readable to me.

```suggestion
    const value = event.target.value;
    if (value) setCount(parseInt(value);
  };
```

  // Validate count to ensure it is within the bounds of the jsonData.shows array

>   const isValidCount =
>     count !== null && count >= 0 && count < jsonData[0].shows.length;

Reviewer: Again, lint with Crawlee ESLint. This definitely doesn't need to be on two lines.

Btw, can `count` really be negative? I guess it can't, right? If not, this would be a better solution:

```suggestion
const isValidCount = count && count < jsonData[0].shows.length;
```

Btw, why `count < jsonData[0].shows.length` ? Shouldn't it be `<=`?

  return (
    <div className="app-container">

>       <h1 className="header">Netflix Web Show Recommender</h1>

Reviewer: Would be cool to split this big wall of HTML into some manageable chunks. Perhaps create some functions to generate this?

                  <li key={index} className="show-item">
                    {show}
                  </li>
                ))}
              </ul>
            </div>
          </>
        )}
      </div>
    </div>
  );
}

export default App;
```

In this code snippet, the `genre` array is used to display the list of genres. User can select their desired genre and based upon that a list of web shows available on Netflix will be displayed using the `shows` array.

Make sure to update CSS on the `App.css` file from here: [https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css](https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css)

and download and save this image file in main project folder: [Download Image](https://raw.githubusercontent.com/ayush2390/web-show-recommender/main/Netflix.png)

Now our project is ready!

## Result

Now, to run your project in the localhost, run this command:

```
npm run dev
```

This command will run your project on the localhost. Here is the demo of the project:

![result](assets/result.gif)

Project link - [https://github.com/ayush2390/web-show-recommender](https://github.com/ayush2390/web-show-recommender)

> In this project, we used Crawlee to scrape Netflix; similarly, Crawlee can be used to scrape Single Application Pages (SPAs) to JavaScript-heavy websites. The best part is all of this can be done while coding in JavaScript/TypeScript and using a single library.

Reviewer: ```suggestion
In this project, we used Crawlee to scrape Netflix; similarly, Crawlee can be used to scrape Single Application Pages (SPAs) and JavaScript-rendered websites. The best part is all of this can be done while coding in JavaScript/TypeScript and using a single library.
```

> If you want to learn more about Crawlee, I recommend going through their [documentation](https://crawlee.dev/docs/quick-start) and the [Crawlee web scraping tutorial by Apify](https://blog.apify.com/crawlee-web-scraping-tutorial/).

Reviewer: ```suggestion
If you want to learn more about Crawlee, I recommend going through the [documentation](https://crawlee.dev/docs/quick-start) and the [Crawlee web scraping tutorial by Apify](https://blog.apify.com/crawlee-web-scraping-tutorial/).
```

# Building a Netflix Web Show Recommender with Crawlee and React

> Creating a web show recommender for Netflix involves scraping data from web pages efficiently. In this article, we'll guide you through the process of using Crawlee. By the end of this guide, you'll have a functional web show recommender that can fetch and scrape Netflix data to provide tailored show suggestions. 

Reviewer 2: ```suggestion
Creating a web show recommender for Netflix involves scraping data from web pages efficiently. In this article, we'll guide you through the process of using Crawlee. By the end of this guide, you'll have a functional web show recommender that can fetch and scrape Netflix data to provide show suggestions. 
```

> Whether you're a developer looking to enhance your scraping skills or a Netflix enthusiast wanting to build a customized recommendation engine, this project has something valuable for you. Let's dive in and explore how Crawlee can help us achieve this.

Reviewer 2: ```suggestion
Whether you're a developer looking to enhance your scraping skills or a Netflix enthusiast wanting to build a show recommendation engine, this project has something valuable for you. Let's dive in and explore how Crawlee can help us achieve this.
```

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Node.js from their [official website](https://nodejs.org/en/).

## Creating React app

Firstly, we will create a React app (for frontend) using Vite. Run this command in the terminal to create react app:

```
npx create-vite@latest
```

You can check out the [Vite Docs](https://vitejs.dev/guide/) to create a React app.

Once the React app is created, open it in VS Code.

![react](assets/react.png)

This will be the structure of your React app.

Run `npm run dev` command in the terminal to run the app.

![viteandreact](assets/viteandreact.png)

This will be the output displayed by the React app.

## Adding Scraper code

As per our project requirements, we will scrape the genres and the titles of the shows available on Netflix.

Let’s start building the scraper code.

### Installation

We will install two libraries - `crawlee` and `cheerio`. Run this command to install these libraries:

```
npm install crawlee cheerio
```

:::note
`cheerio` is not bundled with Crawlee to reduce install size and allow greater flexibility. Therefore we need to install both libraries separately.
:::

> Crawlee utilizes Cheerio for [HTML parsing and scraping](https://crawlee.dev/blog/scrapy-vs-crawlee#html-parsing-and-scraping) of static websites. While faster and less resource-intensive, it can only scrape websites that do not require JavaScript rendering, making it unsuitable for SPAs (single page applications).

Reviewer 2: proof of faster and less resource intensive?

plus worth mentioning Crawlee also supports Playwright and Puppeteer that supports JS-heavy sites too.

Here’s the full code that we will use in our project:

```js
import { CheerioCrawler, log, Dataset } from "crawlee";

const crawler = new CheerioCrawler({
  requestHandler: async ({ request, parseWithCheerio, pushData }) => {
    log.info(`Processing: ${request.url}`);

    // Use parseWithCheerio for efficient HTML parsing
    const $ = await parseWithCheerio();

    // Extract genre titles
    const titles = $(".nm-collections-row-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Extract show titles
    const shows = $(".nm-collections-title-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Prepare data for pushing
    const allShows = [];
    let chunk = [];
    shows.forEach((show) => {
      chunk.push(show);
      if (chunk.length === 40) {
        allShows.push(chunk);
        chunk = [];
      }
    });
    if (chunk.length > 0) {
      allShows.push(chunk);
    }

    await pushData({
      genre: titles,
      shows: allShows,
    });
  },

  // Limit crawls for efficiency
  maxRequestsPerCrawl: 20,
});

await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);
await Dataset.exportToJSON("results");
```

> Now we will run the crawlee to scrape the website. Before running the crawlee, we need to tweak the `package.json` file. We will add the `start` script targeting the `scraper.js` file to run the crawlee.

Reviewer 2: ```suggestion
Now, we will run Crawlee to scrape the website. Before running Crawlee, we need to tweak the `package.json` file. We will add the `start` script targeting the `scraper.js` file to run Crawlee.
```

                  <li key={index} className="show-item">
                    {show}
                  </li>
                ))}
              </ul>
            </div>
          </>
        )}
      </div>
    </div>
  );
}

export default App;
```

In this code snippet, the `genre` array is used to display the list of genres. User can select their desired genre and based upon that a list of web shows available on Netflix will be displayed using the `shows` array.

Make sure to update CSS on the `App.css` file from here: [https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css](https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css)

and download and save this image file in main project folder: [Download Image](https://raw.githubusercontent.com/ayush2390/web-show-recommender/main/Netflix.png)

Now our project is ready!

## Result

Now, to run your project in the localhost, run this command:

```
npm run dev
```

This command will run your project on the localhost. Here is the demo of the project:

![result](assets/result.gif)

Project link - [https://github.com/ayush2390/web-show-recommender](https://github.com/ayush2390/web-show-recommender)

> In this way, you can scrape data from any website using crawlee and create some unique projects using the scraped data.

Reviewer 2: ```suggestion
In this project, we used Crawlee to scrape Netflix; similarly, Crawlee can be used to scrape Single Application Pages (SPAs) to JavaScript-heavy websites. The best part is all of this can be done while coding in JavaScript/TypeScript and using a single library.

If you want to learn more about Crawlee, I recommend going through their [documentation](https://crawlee.dev/docs/quick-start) and the [Crawlee web scraping tutorial by Apify](https://blog.apify.com/crawlee-web-scraping-tutorial/).
```

> Want to know more about Crawlee? You can watch our Crawlee guide video:
> 
> <iframe
>   width="560"
>   height="315"
>   src="https://www.youtube.com/embed/g1Ll9OlFwEQ?si=EMeYil29LV0oDOxc"
>   title="YouTube video player"
>   frameborder="0"
>   allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
>   referrerpolicy="strict-origin-when-cross-origin"
>   allowfullscreen
> ></iframe>

Reviewer 2: ```suggestion
```

# Building a Netflix Web Show Recommender with Crawlee and React

> Creating a personalized web show recommender for Netflix involves scraping data from web pages efficiently. In this article, we'll guide you through the process of using Crawlee. By the end of this guide, you'll have a functional web show recommender that can fetch and analyze Netflix data to provide tailored show suggestions. 

Reviewer: Please be careful with using words like "personalized" (personalized how?), or "tailored" (again, how?)

In developer content, instead of using those "meaningless" adjectives, let's say what the project will do.

Also, how does the project "analyze" data? There's no analytics in place. Don't overpromote. Just describe.

Here’s the full code that we will use in our project:

```js
import { CheerioCrawler, log, Dataset } from "crawlee";

const crawler = new CheerioCrawler({
  requestHandler: async ({ request, parseWithCheerio }) => {
    log.info(`Processing: ${request.url}`);

    // Use parseWithCheerio for efficient HTML parsing
    const $ = await parseWithCheerio();

    // Extract genre titles
    const titles = $(".nm-collections-row-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Extract show titles
    const shows = $(".nm-collections-title-name")
      .map((_, el) => $(el).text().trim())
      .get();

    // Prepare data for pushing
    const allShows = [];
    let chunk = [];
    shows.forEach((show) => {
      chunk.push(show);
      if (chunk.length === 40) {
        allShows.push(chunk);
        chunk = [];
      }
    });
    if (chunk.length > 0) {
      allShows.push(chunk);
    }

    // await Dataset.pushData({ titles, allShows });
    await Dataset.pushData({
      genre: titles,
      shows: allShows,
    });
  },

  // Limit crawls for efficiency
  maxRequestsPerCrawl: 20,
});

await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);

> await Dataset.exportToJSON("results");

Reviewer: I don't think you can delete that. You're importing `Dataset` not `exportToJSON`

> # Building a Netflix Web Show Recommender with Crawlee

Reviewer: I would mention React in the title as well. It could attrack more readers to the article and hence to Crawlee.

```suggestion
# Building a Netflix Web Show Recommender with Crawlee and React
```

Author: Done

> 

Reviewer: We should make it super clear at the beginning that this is a community written tutorial. We want to let the community know that we're ready to showcase their tutorials.

An admonition would be great for that.

We will set up Crawlee and understand it's working. Let’s get started!

## Prerequisites

To use Crawlee, you need to have Node.js 16 or higher version.

You can install the latest version of Node.js from their [official website](https://nodejs.org/en/).

## Creating React app

Firstly, we will create a React app (for frontend) using Vite. Run this command in the terminal to create react app:

```
npx create-vite@latest
```

You can check out the [Vite Docs](https://vitejs.dev/guide/) to create a React app.

Once the React app is created, open it in VS Code.

![react](assets/react.png)

This will be the structure of your React app.

Run `npm run dev` command in the terminal to run the app.

![viteandreact](assets/viteandreact.png)

This will be the output displayed by the React app.

## Adding Scraper code

As per our project requirements, we will scrape the genres and the titles of the shows available on Netflix.

Let’s start building the scraper code.

### Installation

We will install two libraries - `crawlee` and `playwright`. Run this command to install these libraries:

```
npm install crawlee playwright
```

> Note: `playwright` is not bundled with Crawlee to reduce install size and allow greater flexibility. Therefore we need to install both libraries separately.

Reviewer: Can we use the same admonitions in Blog as we use in docs and examples? Aka

```
:::note
some text
:::
```

Instead of simply **Note:**

Author: I have added this in the blog, but it is not giving the desired result. Please have a look at it

> Crawlee, built on Playwright, empowers [headless browser scraping](https://crawlee.dev/blog/scrapy-vs-crawlee#headless-browsing-and-js-rendering) for complex websites. While slower and more resource-intensive, it tackles intricate JavaScript-heavy scenarios that basic crawlers can't handle.

Reviewer: Cut the fluff.

```suggestion
Crawlee utilizes Playwright for [headless browser scraping](https://crawlee.dev/blog/scrapy-vs-crawlee#headless-browsing-and-js-rendering) of JavaScript rendered websites. While slower and more resource-intensive, it can scrape websites that pure HTTP crawlers can't handle, such as SPAs (single page applications).
```

Author: Done

After installing the libraries, it’s time to create the scraper code.

Create a file in `src` directory and name it `scraper.js`. Whole scraper code will be created in this file.

### Importing libraries

`PlaywrightCrawler` class has been imported from the `crawlee` library. `PlaywrightCrawler` crawls the web using a headless browser controlled by the `Playwright` library.

```js
import { PlaywrightCrawler } from "crawlee";
```

> 

Reviewer: Let's remove this section. We don't need to repeat what we already said about headless browsers and importing libraries is so basic that it doesn't deserve its own section.

Author: Done

### Scraping genres

To scrape the genres, we will utilize the [browser Devtools](https://developer.mozilla.org/en-US/docs/Learn/Common`questions/Tools`and`setup/What`are`browser`developer`tools) to identify the tags and CSS selectors targeting the genre elements on the Netflix website.

![genre](assets/genre.png)

Here, we can observe that the name of the genre is captured by the `span` tag having `nm-collections-row-name` class. So we can use the `span.nm-collections-row-name` selector to capture this and similar elements.

```js
const titles = await page.$$eval("span.nm-collections-row-name", (elements) => {
  // Map the text content of each element
  return elements.map((el) => el.textContent.trim());
});

> ```

Reviewer: There's no waiting for elements present. So either this will be flaky (because the data is loaded at a random interval and might not be available, or the crawler does not require the use of Playwright (because the data is always there in the initial HTML). Which is it?

Note that I'm not asking to simply add a wait. It's important to KNOW whether the dynamic rendering is even needed here, because we don't want to be showcasing a suboptimal way of making a crawler.

Author: Done

Reviewer 2: @ayush2390 This is not addressed yet; what Ondra is trying to say is that maybe the website is not JS heavy at all or the data that we are scraping is present in the initial HTML of the page, and we don't need a playwright there, which is true, and I checked the code running it with `CheerioCrawler` and it was working, which means that data is present in the initial html.

so maybe we need to scrape some other website that actually needs to be crawled with playwright to show the optimal way of doing it.

This will give us the list of all the genres stored in the `titles` array.

### Scraping shows

To scrape the titles of shows, we will use the same method we used for scraping genres. Let’s find out the tag capturing the shows’ names.

![title](assets/title.png)

Here, we can observe that the title of the show is captured by the `span` tag having `nm-collections-title-name` class. So we can use the `span.nm-collections-title-name` selector to capture this and similar elements.

```js

> const shows = await page.$$eval(

Reviewer: Instead of using 2x `page.$$eval()`, it might be better to use `parseWithCheerio()`

Author: Done

  "span.nm-collections-title-name",
  (elements) => {
    // Map the text content of each element
    return elements.map((el) => el.textContent.trim());
  }
);

> ```

Reviewer: This piece of code is formatted differently than the above one. Please choose one formatting style.

Personally, I would choose the former one. This is too narrow for my taste and is not in line with how we format code in Crawlee docs and examples.

Author: Done

This will scrape the title of all the shows and store it in the `shows` array.

```js
let allShows = [];
let totalShows = [];
shows.map((show) => {
  if (allShows.length == 40) {
    totalShows.push(allShows);
    allShows = [];
  }
  allShows.push(show);
});
```

> Here, we have performed some basic operations on the `shows` array to get the `totalShows` array in the desired format.

Reviewer: From a quick look I don't understand what this is good for. Please either simplify it or explain why this is needed instead of just saying "we have performed basic operations". I can see that, but why? What is the desired format and why is it desired?

### Storing data

Now we have every data that we want for our project and it’s time to store or save the scraped data. To store the data, crawlee comes with a `pushData()` method.

The [pushData()](https://crawlee.dev/docs/introduction/saving-data) method creates a storage folder in the project directory and stores the scraped data in the json format.

```js
await pushData({
  genre: titles,
  shows: totalShows,
});
```

This will save the `titles` and `totalShows` arrays as values in the `genre` and `shows` keys.

Here’s the full code that we will use in our project:

```js
import { PlaywrightCrawler } from "crawlee";

const crawler = new PlaywrightCrawler({

>   requestHandler: async ({ page, request, enqueueLinks, pushData }) => {

Reviewer: I see `enqueueLinks` here, but I don't see it used anywhere. Do we scrape just one page?

Author: Done

    console.log(`Processing: ${request.url}`);

>     console.log(request.label);

Reviewer: I recommend using Crawlee's internal logging with `log.info`

      "span.nm-collections-row-name",
      (elements) => {
        // Map the text content of each element
        return elements.map((el) => el.textContent.trim());
      }
    );
    // console.log(titles);
    const shows = await page.$$eval(
      "span.nm-collections-title-name",
      (elements) => {
        // Map the text content of each element
        return elements.map((el) => el.textContent.trim());
      }
    );

    let allShows = [];
    let totalShows = [];
    shows.map((show) => {
      if (allShows.length == 40) {
        totalShows.push(allShows);
        allShows = [];
      }
      allShows.push(show);
    });

    await pushData({
      genre: titles,
      shows: totalShows,
    });
  },

  // Let's limit our crawls to make our tests shorter and safer.
  maxRequestsPerCrawl: 20,
});

await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);
```

The `maxRequestPerCrawl` limits the crawlee to crawl only the first page of the given URL to make the tests shorter and safer. You can ignore this code line to crawl the entire website.

```js
await crawler.run(["https://www.netflix.com/in/browse/genre/1191605"]);
```

> In this code snippet, we have defined the URL that we want to scrape data from. You can scrape multiple URLs also by adding the other URLs in the array.

Reviewer: It's not ideal to explain a piece of code after it's been used in a full code example. Especially when there's another `maxRequestsPerCrawl` explanation between them.

Let's just use comments inside the code. The function is quite self explanatory.

Author: Addressed this comment

Now we will run the crawlee to scrape the website. Before running the crawlee, we need to tweak the `package.json` file. We will add the `start` script targeting the `scraper.js` file to run the crawlee.

Add the following code in `"scripts"` object:

```
"start": "node src/scraper.js"
```

and save it. Now run this command to run the crawlee the scrape the data:

```
npm start
```

After running this command, you will see a `storage` folder having `datasets\default\000000001.json` file. In this json file, the scrapped data will be stored in json format.

Now we can use this json data and display it in the `App.jsx` component to create the project.

In the `App.jsx` component, we will import `jsonData` from the `000000001.json` file:

```js
import jsonData from "../storage/datasets/default/000000001.json";

> ```

Reviewer: Instead of pulling the data from storage, it would be better to call [`Dataset.exportToJSON()`](https://crawlee.dev/api/core/class/Dataset#exportToJSON) after the crawl to merge all results and save them into a file with a nice name and then load from this file.

Author: I have done it and used the `Dataset.exportToJSON()` in the code

          {jsonData.genre.map((genre, key) => {
            return (
              <option key={key} value={key}>
                {genre}
              </option>
            );
          })}
        </select>
      </div>
      <div className="shows-container">
        {count && (
          <>
            <div className="shows-list">
              <ul>
                {jsonData.shows[count].slice(0, 20).map((show, index) => (
                  <li key={index} className="show-item">
                    {show}
                  </li>
                ))}
              </ul>
            </div>
            <div className="shows-list">
              <ul>
                {jsonData.shows[count].slice(20).map((show, index) => (
                  <li key={index} className="show-item">
                    {show}
                  </li>
                ))}
              </ul>
            </div>
          </>
        )}
      </div>
    </div>
  );
}

export default App;
```

In this code snippet, the `genre` array is used to display the list of genres. User can select their desired genre and based upon that a list of web shows available on Netflix will be displayed using the `shows` array.

> Make sure to update CSS on the `App.css` file from here: [https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css](https://github.com/ayush2390/web-show-recommender/blob/main/src/App.css)

Reviewer: Did you create the CSS yourself or is it extracted from some other tutorial or project? Open source or not, we should make sure that all of the content here is either original, or that we at least mention the original authors. It's extremely important not to plagiarize stuff.

Author: Yes, I have created the CSS by myself. It's fully created by me
